### PR TITLE
Upgrade okhttp3 and okio

### PR DESCRIPTION
**Pull request by ajayk**

### CODE REVIEW

File: src/main/java/io/minio/MinioClient.java [hunk 1]
```diff
@@ -2572,11 +2574,7 @@ private void completeMultipart(String bucketName, String objectName, String uplo
       if (scanner.hasNext()) {
         bodyContent = scanner.next();
       }
-    } catch (EOFException e) {
```
Comment (ajayk) [inline src/main/java/io/minio/MinioClient.java]: The EOFException is never thrown with the latest libraries

**Comment by harshavardhana** [comment]
Is there a reason to update okhttp here? any benefits / pitfalls?

**Comment by ajayk** [comment]
Nothing specific though , OkHttpClient is now stateless and has a builder interfaces which is a cleaner implementation than the version 2 , Apart from the Numerous fixes , HttpLoggingInterceptor can be used in the clients which can help in debugging the network call stacks

**Comment by harshavardhana** [comment]
> Nothing specific though , OkHttpClient is now stateless and has a builder interfaces which is a cleaner implementation than the version 2 , Apart from the Numerous fixes , HttpLoggingInterceptor can be used in the clients which can help in debugging the network call stacks , CertificatePinner has some fixes as well .

Okay cool.

**Comment by harshavardhana** [comment]
> The EOFException is never thrown with the latest libraries ,i have updated the tests as well , Can you please check if this is the intended behavior

@balamurugana can you check?

File: src/main/java/io/minio/MinioClient.java (hunk 1 above)
Comment (balamurugana) [inline src/main/java/io/minio/MinioClient.java]: I see the change from EOFException to ConnectException.  Is it required to catch ConnectException and handle here?

**Comment by balamurugana** [comment]
@ajayk can you fix checkstyle errors https://travis-ci.org/minio/minio-java/jobs/151411663 ?

**Comment by balamurugana** [comment]
> The EOFException is never thrown with the latest libraries ,i have updated the tests as well , Can you please check if this is the intended behavior
> 
> @balamurugana can you check?

I am checking now.

**Comment by ajayk** [comment]
@balamurugana  will fix them soon

**Comment by balamurugana** [comment]
> @balamurugana will fix them soon

Thanks for fixing it :)

**Comment by balamurugana** [comment]
@ajayk `./gradlew build`  fails with below error

```
:test

io.minio.MinioClientTest > testPutIncompleteSmallPut FAILED
    java.lang.Exception: Unexpected exception, expected<java.net.ConnectException> but was<java.io.EOFException>

        Caused by:
        java.io.EOFException
            at okio.RealBufferedSink.write(RealBufferedSink.java:107)
            at io.minio.MinioClient$1.writeTo(MinioClient.java:661)
            at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:47)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
            at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:109)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
            at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:124)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
            at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:170)
            at okhttp3.RealCall.execute(RealCall.java:60)
            at io.minio.MinioClient.execute(MinioClient.java:803)
            at io.minio.MinioClient.executePut(MinioClient.java:1060)
            at io.minio.MinioClient.executePut(MinioClient.java:1084)
            at io.minio.MinioClient.putObject(MinioClient.java:2044)
            at io.minio.MinioClient.putObject(MinioClient.java:2066)
            at io.minio.MinioClient.putObject(MinioClient.java:2018)
            at io.minio.MinioClientTest.testPutIncompleteSmallPut(MinioClientTest.java:539)

32 tests completed, 1 failed
:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/bala/java/src/github.com/minio-java/build/reports/tests/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 1 mins 28.605 secs
```

This needs to be fixed.

Not sure why travis-ci is not getting this.

**Comment by ajayk** [comment]
@balamurugana i didn't get this test failures on my local machine as well ,  i ran the tests couple of times , so i commented on the behavior

**Comment by balamurugana** [comment]
> @balamurugana i didn't get this test failures on my local machine as well , i ran the tests couple of times , so i commented on the behavior

@ajayk I am not able to reproduce the error in new machine, but getting it always in my laptop.

It seems the exception `EOFException` I get in my laptop looks correct to me as we read beyond available data.  `ConnectException` for this case is quite strange to me.

Till I complete the debugging, please rebase your patch.

**Comment by balamurugana** [comment]
@ajayk 

As per `ConnectException` trace, its something fishy.

```
io.minio.MinioClientTest > testPutIncompleteSmallPut FAILED
    java.lang.Exception: Unexpected exception, expected<io.minio.errors.MinioException> but was<java.net.ConnectException>

        Caused by:
        java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:42350
            at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:189)
            at okhttp3.internal.connection.RealConnection.buildConnection(RealConnection.java:173)
            at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:114)
            at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:193)
            at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:129)
            at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:98)
            at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
            at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:109)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
            at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:124)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
            at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
            at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:170)
            at okhttp3.RealCall.execute(RealCall.java:60)
            at io.minio.MinioClient.execute(MinioClient.java:803)
            at io.minio.MinioClient.executePut(MinioClient.java:1060)
            at io.minio.MinioClient.executePut(MinioClient.java:1084)
            at io.minio.MinioClient.putObject(MinioClient.java:2044)
            at io.minio.MinioClient.putObject(MinioClient.java:2066)
            at io.minio.MinioClient.putObject(MinioClient.java:2018)
            at io.minio.MinioClientTest.testPutIncompleteSmallPut(MinioClientTest.java:539)

32 tests completed, 1 failed
:test FAILED
```

The exception is clear to say `Failed to connect to localhost/0:0:0:0:0:0:0:1:42350`.  Can you check why we are not able to connect to the mock server?  May be related to IPv6?

**Comment by balamurugana** [comment]
I found the problem.

If `localhost` name is set to IPv6 in /etc/hosts file, okhttp fails to connect.  Below is the respective /etc/hosts entry.
```
::1     localhost ip6-localhost ip6-loopback
```

In my laptop, it doesn't point localhost to IPv6 by default.  Need to check further.

**Comment by balamurugana** [comment]
Refer https://github.com/square/okhttp/issues/1069

**Comment by balamurugana** [comment]
@ajayk I see mock server listens to IPv6 address. There is a possibility that the client is trying to connect using IPv4 and getting `ConnectException`.  Please check this.

**Comment by ajayk** [comment]
@balamurugana Thanks , i will have a look into it

**Comment by sachinwalia2k8** [comment]
@ajayk any update if okhttp3 can be merged with minio-java?

**Comment by ajayk** [comment]
@sachinwalia2k8 we have one testcase which is failing this is specifically in the mockserver 
@balamurugana i tried to run the test using the minio-server and s3 endpoints using the minio-java client [the okhttp3 branch and the previous stable release 2.0.1 , The results were consistent across , i don't think there is any change in the behavior ,  I am consistently getting InsufficientDataException in both the versions against a minio server and s3 endpoints

o.minio.MinioClientTest > testPutIncompleteSmallPut FAILED
    java.lang.Exception: Unexpected exception, expected<java.net.ConnectException> but was<io.minio.errors.InsufficientDataException>

        Caused by:
        io.minio.errors.InsufficientDataException: Insufficient data.  bytes read 10 expected 11
            at io.minio.Digest.readBytes(Digest.java:220)
            at io.minio.Digest.md5Hash(Digest.java:117)
            at io.minio.MinioClient.createRequest(MinioClient.java:715)
            at io.minio.MinioClient.execute(MinioClient.java:783)
            at io.minio.MinioClient.executePut(MinioClient.java:1060)
            at io.minio.MinioClient.executePut(MinioClient.java:1084)
            at io.minio.MinioClient.putObject(MinioClient.java:2044)
            at io.minio.MinioClient.putObject(MinioClient.java:2066)
            at io.minio.MinioClient.putObject(MinioClient.java:2018)
            at io.minio.MinioClientTest.testPutIncompleteSmallPut(MinioClientTest.java:542)


Results using the Aws-SDK-Java s3client 

`
Caught an AmazonClientException, which means the client encountered a serious internal problem while trying to communicate with S3, such as not being able to access the network.
Error Message: Data read has a different length than the expected: dataLength=10; expectedLength=11; includeSkipped=false; in.getClass()=class com.amazonaws.internal.ReleasableInputStream; markedSupported=true; marked=0; resetSinceLastMarked=false; markCount=1; resetCount=0
`

**Comment by balamurugana** [comment]
@ajayk Have you run unit tests with IPv6 setup for localhost I mentioned above?

You had changed expected exception from `EOFException` to `ConnectException` in `io.minio.MinioClientTest.testPutIncompleteSmallPut()`.

This is not the right fix.

**Comment by ajayk** [comment]
@balamurugana keep the PR onhold until i figure out the connectionException , Thanks for your patience

**Comment by harshavardhana** [comment]
Deferred and closed for now, we will address this in next major update. Thanks for the fix @ajayk

**Pull request closed without merging**